### PR TITLE
Port dirty rect handling from JGRPP

### PR DESCRIPTION
**Pull request by ldpl**
## Motivation / Problem

I didn't want to deal with vehicles and stuff constantly redrawing minimap in cmclient so I yoinked dirty rect handling from JGRPP. And thought I might as well PR it for vanilla since I've already did the work of isolating it from other changes.

But other that that I'm not sure where this PR is going to go. I didn't do any performance testing or code cleanup. And not even sure if I'm ever going to do that, So if anyone wants to continue working on this feel free to do so.

## Description

Combines these commits and maybe some more:
* Change how dirty screen, window and viewport areas are tracked for later redrawing: https://github.com/JGRennison/OpenTTD-patches/commit/8f442500eac717a4ba84814972680740709e32d3
* Fix window re-rendering when ReInit() called within OnPaint/draw handler:  https://github.com/JGRennison/OpenTTD-patches/commit/c227d30801ab91a47e15917fc73da3a6aca9cdc0
* Add gfx_debug console command: https://github.com/JGRennison/OpenTTD-patches/commit/d29a08dfe10e4a88e31e7966e3748e9f8adcbf13
* Use new/delete for Viewport info struct: https://github.com/JGRennison/OpenTTD-patches/commit/4bad77cdf01b6be499b17fb27d78fa27f31d53f3

## Limitations

* Didn't test performance
* Didn't try to understand what most of the code does.
* Combined it from bits and pieces, may have missed something.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

**Comment by JGRennison** [comment]
You might want https://github.com/JGRennison/OpenTTD-patches/commit/8071976b0ec607cc379e210cfe570f769f9f3645 https://github.com/JGRennison/OpenTTD-patches/commit/3e017ddaa9e043f5f959affe77346e65baa71114 https://github.com/JGRennison/OpenTTD-patches/commit/6ecf232edd296cdc22667fd58e45e14fefabbba7 https://github.com/JGRennison/OpenTTD-patches/commit/0a6a38886d197fd1d4e631039c3ad0eeac896c15 https://github.com/JGRennison/OpenTTD-patches/commit/5ac240147eaa07764b8ae896a14cd084d156a8d8 as well.

**Comment by ldpl** [comment]
Added them, ty.

**Comment by ldpl** [comment]
Added https://github.com/JGRennison/OpenTTD-patches/commit/760a1f0163df819d7e7ffbdc5acfc0077d95dcfe
Otherwise minimizing vehicle window crashed the game.

File: src/viewport.cpp [hunk 1]
```diff
@@ -1906,6 +1930,32 @@ void UpdateViewportPosition(Window *w)
...
+	// 	} else {
+	// 		vp->land_pixel_cache.assign(vp->ScreenArea(), 0xD7);
+	// 	}
+	// } else {
+		// vp->map_draw_vehicles_cache.vehicle_pixels.clear();
+		// vp->land_pixel_cache.clear();
+		// vp->land_pixel_cache.shrink_to_fit();
+	// }
+	// vp->update_vehicles = true;
+	// FillViewportCoverageRect();
```
Comment (LordAro) [inline src/viewport.cpp:1956]: commented code

File: src/viewport.cpp [hunk 2]
```diff
@@ -1936,12 +1986,40 @@ static bool MarkViewportDirty(const Viewport *vp, int left, int top, int right,
...
-		UnScaleByZoom(bottom, vp->zoom) + vp->top + 1
-	);
+	uint x = std::max<int>(0, UnScaleByZoomLower(left, vp->zoom) - vp->dirty_block_left_margin) >> vp->GetDirtyBlockWidthShift();
+	uint y = UnScaleByZoomLower(top, vp->zoom) >> vp->GetDirtyBlockHeightShift();
+	uint w = (std::max<int>(0, UnScaleByZoomLower(right, vp->zoom) - 1 - vp->dirty_block_left_margin) >> vp->GetDirtyBlockWidthShift()) + 1 - x;
+	uint h = ((UnScaleByZoom(bottom, vp->zoom) - 1) >> vp->GetDirtyBlockHeightShift()) + 1 - y;
+
+	// TODO somehow JGRPP avoids these checks
+	if (x >= vp->dirty_blocks_per_row) return false;
+	if (y >= vp->dirty_blocks_per_column) return false;
```
Comment (LordAro) [inline src/viewport.cpp:1996]: TODO

File: src/viewport.cpp [hunk 3]
```diff
@@ -1936,12 +1986,40 @@ static bool MarkViewportDirty(const Viewport *vp, int left, int top, int right,
...
+		uint t = UnScaleByZoomLower(top, vp->zoom);
+		uint w = UnScaleByZoom(right, vp->zoom) - l;
+		uint h = UnScaleByZoom(bottom, vp->zoom) - t;
+		uint bitdepth = BlitterFactory::GetCurrentBlitter()->GetScreenDepth() / 8;
+		uint8 *land_cache = vp->land_pixel_cache.data() + ((l + (t * vp->width)) * bitdepth);
+		while (--h) {
+			memset(land_cache, 0xD7, (size_t)w * bitdepth);
+			land_cache += vp->width * bitdepth;
+		}
+	}*/
```
Comment (LordAro) [inline src/viewport.cpp:2022]: commented code

File: src/viewport_type.h [hunk 4]
```diff
@@ -32,6 +34,35 @@ struct Viewport {
...
+			this->dirty_blocks.assign(this->dirty_blocks.size(), false);
+			this->is_dirty = false;
+		}
+		this->is_drawn = false;
+	}
+
+private:
+	uint GetDirtyBlockShift() const
+	{
+		// if (this->zoom >= ZOOM_LVL_DRAW_MAP) return 3;
```
Comment (LordAro) [inline src/viewport_type.h:62]: commented code

File: src/viewport.cpp [hunk 5]
```diff
@@ -335,6 +338,15 @@ static void DoSetViewportPosition(Window::IteratorToFront it, int left, int top,
 	}
 }
 
+inline void UpdateViewportDirtyBlockLeftMargin(Viewport *vp)
+{
+	// if (vp->zoom >= ZOOM_LVL_DRAW_MAP) {
+	// 	vp->dirty_block_left_margin = 0;
+	// } else {
+		vp->dirty_block_left_margin = UnScaleByZoomLower((-vp->virtual_left) & 127, vp->zoom);
+	// }
```
Comment (LordAro) [inline src/viewport.cpp:347]: commented code

File: src/gfx.cpp [hunk 6]
```diff
@@ -1627,6 +1630,72 @@ void RedrawScreenRect(int left, int top, int right, int bottom)
...
+	for(; occlusion < _dirty_viewport_occlusions.size(); occlusion++) {
+		const Rect &occ = _dirty_viewport_occlusions[occlusion];
+		if (right > occ.left &&
+				bottom > occ.top &&
+				left < occ.right &&
+				top < occ.bottom) {
+			/* occlusion and draw rectangle intersect with each other */
+			int x;
+
+			if (left < (x = occ.left)) {
```
Comment (LordAro) [inline src/gfx.cpp:1648]: can't say i'm a fan of this. Why not just use occ.foo in the subsequent calls?

File: src/gfx_func.h [hunk 7]
```diff
@@ -21,7 +21,7 @@
  * (this is btw. also possible if needed). This is used to avoid a
  * flickering of the screen by the video driver constantly repainting it.
  *
- * This whole mechanism is controlled by an rectangle defined in #_invalid_rect. This
+ * This whole mechanism was controlled by an rectangle defined in #_invalid_rect. This
```
Comment (LordAro) [inline src/gfx_func.h:24]: I'd imagine this comment block needs properly updating

File: src/window.cpp [hunk 8]
```diff
@@ -2120,7 +2123,14 @@ void ResizeWindow(Window *w, int delta_x, int delta_y, bool clamp_to_screen)
 
 	/* Always call OnResize to make sure everything is initialised correctly if it needs to be. */
 	w->OnResize();
-	w->SetDirty();
+
+	// TODO incapsulate into a function
```
Comment (LordAro) [inline src/window.cpp:2127]: TODO

**Review by LordAro — COMMENTED** [review]
Could do with a load more comments explaining what's going on, but on the whole it looks pretty good

I did some testing with wentbourne - fully zoomed out on a vaguely central city resulted in a game render drop from 130ms -> 105ms, which would suggest a 20% improvement. Not bad.

**Comment by ldpl** [comment]
Well, as I already said in the description I only made this PR because I already had the change isolated from the rest of the code. I didn't even dig too deep into the code to understand it and just added some checks so it doesn't crash. Bringing this to mergeable state is surely going to take more work that I have no plans on doing anytime soon so if you or someone else wants to finish this PR feel free to take it over.

**Comment by 2TallTyler** [comment]
Thanks for sharing your work to pick this out of JGRPP. It doesn't seem like anyone is interested in continuing it right now. I'm going to close this PR, but the code will still be available if anyone wants to pick it up in the future.

**Pull request closed without merging**